File: behavioral/mediator/python/mediator.py

# Mediator interface
class Mediator:
    def notify(self, sender, event_code):
        pass

# Concrete Mediator class
class ConcreteMediator(Mediator):
    def __init__(self, component1, component2):
        self._component1 = component1
        self._component1.set_mediator(self)
        self._component2 = component2
        self._component2.set_mediator(self)

    def notify(self, sender, event_code):
        if event_code == "A":
            print("Mediator reacts to A and triggers following operations:")
            self._component2.do_action2()
        elif event_code == "B":
            print("Mediator reacts to B and triggers following operations:")
            self._component1.do_action1()

# Base Component class
class Component:
    def __init__(self, mediator=None):
        self._mediator = mediator

    def set_mediator(self, mediator):
        self._mediator = mediator

# Component 1
class Component1(Component):
    def do_action1(self):
        print("Component 1 does Action 1.")
        self._mediator.notify(self, "A")

# Component 2
class Component2(Component):
    def do_action2(self):
        print("Component 2 does Action 2.")
        self._mediator.notify(self, "B")

if __name__ == "__main__":
    component1 = Component1()
    component2 = Component2()
    mediator = ConcreteMediator(component1, component2)

    print("Component 1 triggers action A.")
    component1.do_action1()

    print("Component 2 triggers action B.")
    component2.do_action2()

File: behavioral/mediator/csharp/Mediator.cs

using System;

// Mediator interface
public interface IMediator
{
    void Notify(object sender, string eventCode);
}

// Concrete Mediator class
public class ConcreteMediator : IMediator
{
    private Component1 _component1;
    private Component2 _component2;

    public ConcreteMediator(Component1 component1, Component2 component2)
    {
        _component1 = component1;
        _component1.SetMediator(this);
        _component2 = component2;
        _component2.SetMediator(this);
    }

    public void Notify(object sender, string eventCode)
    {
        if (eventCode == "A")
        {
            Console.WriteLine("Mediator reacts to A and triggers following operations:");
            _component2.DoAction2();
        }
        else if (eventCode == "B")
        {
            Console.WriteLine("Mediator reacts to B and triggers following operations:");
            _component1.DoAction1();
        }
    }
}

// Component base class
public abstract class Component
{
    protected IMediator _mediator;

    public Component(IMediator mediator = null)
    {
        _mediator = mediator;
    }

    public void SetMediator(IMediator mediator)
    {
        _mediator = mediator;
    }
}

// Component 1
public class Component1 : Component
{
    public void DoAction1()
    {
        Console.WriteLine("Component 1 does Action 1.");
        _mediator.Notify(this, "A");
    }
}

// Component 2
public class Component2 : Component
{
    public void DoAction2()
    {
        Console.WriteLine("Component 2 does Action 2.");
        _mediator.Notify(this, "B");
    }
}

class Program
{
    static void Main()
    {
        Component1 component1 = new Component1();
        Component2 component2 = new Component2();
        ConcreteMediator mediator = new ConcreteMediator(component1, component2);

        Console.WriteLine("Component 1 triggers action A.");
        component1.DoAction1();

        Console.WriteLine("Component 2 triggers action B.");
        component2.DoAction2();
    }
}
